astropy/coordinates/builtin_frames/icrs_observed_transforms.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Contains the transformation functions for getting to "observed" systems from ICRS.
+Currently that just means AltAz.
+"""
+
+import numpy as np
+import erfa
+
+from astropy import units as u
+from astropy.coordinates.builtin_frames.utils import atciqz, aticq
+from astropy.coordinates.baseframe import frame_transform_graph
+from astropy.coordinates.transformations import FunctionTransformWithFiniteDifference
+from astropy.coordinates.representation import (SphericalRepresentation,
+                                                CartesianRepresentation,
+                                                UnitSphericalRepresentation)
+
+from .icrs import ICRS
+from .altaz import AltAz
+from .utils import PIOVER2
+from ..erfa_astrom import erfa_astrom
+
+
+@frame_transform_graph.transform(FunctionTransformWithFiniteDifference, ICRS, AltAz)
+def icrs_to_altaz(icrs_coo, altaz_frame):
+    # if the data are UnitSphericalRepresentation, we can skip the distance calculations
+    is_unitspherical = (isinstance(icrs_coo.data, UnitSphericalRepresentation) or
+                        icrs_coo.cartesian.x.unit == u.one)
+    # first set up the astrometry context for ICRS<->AltAz
+    astrom = erfa_astrom.get().apco(altaz_frame)
+
+    # correct for parallax to find BCRS direction from observer (as in erfa.pmpx)
+    if is_unitspherical:
+        srepr = icrs_coo.spherical
+    else:
+        observer_icrs = CartesianRepresentation(astrom['eb'] * u.au, xyz_axis=-1)
+        srepr = (icrs_coo.cartesian - observer_icrs).represent_as(
+            SphericalRepresentation)
+
+    # convert to topocentric CIRS
+    cirs_ra, cirs_dec = atciqz(srepr, astrom)
+
+    # now perform AltAz conversion
+    az, zen, ha, odec, ora = erfa.atioq(cirs_ra, cirs_dec, astrom)
+    alt = np.pi/2-zen
+    aa_srepr = SphericalRepresentation(az*u.radian, alt*u.radian, srepr.distance)
```

Comment:
Replace `*` with `<<` and add `copy=False` for avoiding two more copies (all entries are newly created and will be discarded otherwise, so it is safe).

Reply:
👍